### ISSUE TITLE
Better handling when attempting to search with source filtering against docs with disabled source

### ISSUE THREAD

**Issue by ppf2** [elastic/elasticsearch#7758]
Reproducible using Java API 1.3.x.  See repro at :  https://gist.github.com/ppf2/1a14a31b3add6c450c10

In short, if you have a doc type with source disabled and if you attempt to use the Java API to search for docs of this type specifying source filtering via SearchSourceBuilder, you will receive an unintuitive exception:

```java
Exception in thread "main" org.elasticsearch.action.search.SearchPhaseExecutionException: Failed to execute phase [query_fetch], all shards failed; shardFailures {[bqSWDvHIT7WCtEJKymhnHA][disable_source][0]: ElasticsearchIllegalArgumentException[No matching content type for null]}
```

**Comment by cristiboariu** [comment]
+1

**Comment by qwerty4030** [comment]
On latest master this causes an NPE. I'll take a look, it seems like a similar fix to #18957.
```
PUT disable_source
{
  "settings": {
    "number_of_replicas": 0,
    "number_of_shards": 1
  },
  "mappings": {
    "type": {
      "_source": {
        "enabled": false
      }
    }
  }
}

POST disable_source/type/1
{
  "text1": "hello1",
  "text2": "hello2"
}

POST disable_source/_search
{
  "_source": [
    "text1",
    "text2"
  ]
}
```
NPE response:
```
{
   "error": {
      "root_cause": [
         {
            "type": "null_pointer_exception",
            "reason": null
         }
      ],
      "type": "search_phase_execution_exception",
      "reason": "all shards failed",
      "phase": "query_fetch",
      "grouped": true,
      "failed_shards": [
         {
            "shard": 0,
            "index": "disable_source",
            "node": "71DMW_mRRzKDjQFNDpU4Iw",
            "reason": {
               "type": "null_pointer_exception",
               "reason": null
            }
         }
      ],
      "caused_by": {
         "type": "null_pointer_exception",
         "reason": null
      }
   },
   "status": 500
}
```
